### PR TITLE
[SNAP-3459] - Do not rely on EPSG and check the actual axis name order

### PR DESCRIPTION
**Pull request by oanahogoiu**

### CODE REVIEW

File: snap-jp2-reader/src/main/java/org/esa/snap/jp2/reader/metadata/Jp2XmlMetadata.java [hunk 1]
```diff
@@ -214,4 +225,15 @@ public List<Point2D> getPolygonPositions() {
...
     }
+
+    /**
+     * SNAP-3459 (https://senbox.atlassian.net/browse/SNAP-3459)
+     * Do not rely on EPSG and check the actual axis name order
+     * @return true is the axis order is reversed (e.g. [y,x] instead of [x,y])
+     */
+    private boolean isReversedAxisOrder() {
+        String[] axisName = getAttributeValues(JP2ProductReaderConstants.TAG_AXIS_NAME);
+        return (axisName[0].contains("y") || axisName[0].contains("lat")) ? true : false;
```
Comment (marpet) [inline snap-jp2-reader/src/main/java/org/esa/snap/jp2/reader/metadata/Jp2XmlMetadata.java:236]: can be simplified to:  
`return axisName[0].contains("y") || axisName[0].contains("lat");`

File: snap-jp2-reader/src/main/java/org/esa/snap/jp2/reader/metadata/Jp2XmlMetadata.java [hunk 2]
```diff
@@ -134,15 +134,18 @@ public String getProductDescription() {
         return null;
     }
 
+    /**
```
Comment (marpet) [inline snap-jp2-reader/src/main/java/org/esa/snap/jp2/reader/metadata/Jp2XmlMetadata.java:137]: Don't use api-doc comments for code comments. Simply start with /* instead with /**

Same for the other 3 comments

File: snap-jp2-reader/src/main/java/org/esa/snap/jp2/reader/metadata/Jp2XmlMetadata.java [hunk 3]
```diff
@@ -134,15 +134,18 @@ public String getProductDescription() {
...
     public Point2D getOrigin() {
         Point2D origin = null;
         String coords = getAttributeValue(JP2ProductReaderConstants.TAG_ORIGIN, null);
         if (coords != null) {
-            // SNAP-1400 (https://senbox.atlassian.net/browse/SNAP-1400)
-            // origin = new Point2D.Double(Double.parseDouble(coords.split(" ")[0]), Double.parseDouble(coords.split(" ")[1]));
-            // there are JP2 files with values like "  42.788298611111109   23.271622685185189", resulting in empty tokens and tokens starting with white space
             final String[] result = Arrays.stream(coords.split(" ")).filter(p -> !StringUtils.isEmpty(p)).map(String::trim).toArray(String[]::new);
-            if (!getCrsGeocoding().equals("EPSG::4326")) {
+            if (!isReversedAxisOrder()) {
```
Comment (marpet) [inline snap-jp2-reader/src/main/java/org/esa/snap/jp2/reader/metadata/Jp2XmlMetadata.java:148]: Not using the inversion here saves one operation also one inversion less the brain needs to follow.
Same for the other usages of isReversedAxisOrder()

File: snap-jp2-reader/src/main/java/org/esa/snap/jp2/reader/metadata/Jp2XmlMetadata.java [hunk 4]
```diff
@@ -152,38 +155,46 @@ public Point2D getOrigin() {
...
         String[] values = getAttributeValues(JP2ProductReaderConstants.TAG_OFFSET_VECTOR);
+        String[] result;
         if (values != null) {
-            // SNAP-1400 (https://senbox.atlassian.net/browse/SNAP-1400)
-            //return Double.parseDouble(values[1].split(" ")[1]);
-            // there are JP2 files with values like ["  -0.000004629629630 0", "0    0.000004629629630"]
-            if (!getCrsGeocoding().equals("EPSG::4326")) {
-                final String[] result = Arrays.stream(values[1].split(" ")).filter(p -> !StringUtils.isEmpty(p)).map(String::trim).toArray(String[]::new);
-                return Double.parseDouble(result[1]);
+            if (!isReversedAxisOrder()) {
```
Comment (marpet) [inline snap-jp2-reader/src/main/java/org/esa/snap/jp2/reader/metadata/Jp2XmlMetadata.java:191]: This if-else can be replaced by:
```
            int index = isReversedAxisOrder() ? 0 : 1;
            result = Arrays.stream(values[index].split(" ")).filter(p -> !StringUtils.isEmpty(p)).map(String::trim).toArray(String[]::new);
```
Looks cleaner and is easier to read in my opinion. The other locations can be enahnced similarly.

**Review by marpet — CHANGES_REQUESTED** [review]
Please see the comment added to the files.

Also a few unit level tests would be good for the parsing methods.

File: snap-jp2-reader/src/main/java/org/esa/snap/jp2/reader/metadata/Jp2XmlMetadata.java (hunk 2 above)
Comment (oanahogoiu) [reply snap-jp2-reader/src/main/java/org/esa/snap/jp2/reader/metadata/Jp2XmlMetadata.java]: Code refactoring done.

File: snap-jp2-reader/src/main/java/org/esa/snap/jp2/reader/metadata/Jp2XmlMetadata.java (hunk 3 above)
Comment (oanahogoiu) [reply snap-jp2-reader/src/main/java/org/esa/snap/jp2/reader/metadata/Jp2XmlMetadata.java:148]: Actually, the intention was to have the general use-case (usual axis order) the first one, and then the reversed axis order use-case.

File: snap-jp2-reader/src/main/java/org/esa/snap/jp2/reader/metadata/Jp2XmlMetadata.java (hunk 4 above)
Comment (oanahogoiu) [reply snap-jp2-reader/src/main/java/org/esa/snap/jp2/reader/metadata/Jp2XmlMetadata.java]: Code refactoring done.

File: snap-jp2-reader/src/main/java/org/esa/snap/jp2/reader/metadata/Jp2XmlMetadata.java (hunk 3 above)
Comment (marpet) [reply snap-jp2-reader/src/main/java/org/esa/snap/jp2/reader/metadata/Jp2XmlMetadata.java:148]: Okay, I understand the intention.
In the end I would still prefer:
```
            String x = !isReversedAxisOrder() ? result[0] : result[1];
            String y = !isReversedAxisOrder() ? result[1] : result[0];
            origin = new Point2D.Double(Double.parseDouble(x), Double.parseDouble(y));
```
over 
```
            if (!isReversedAxisOrder()) {
                // the usual order
                origin = new Point2D.Double(Double.parseDouble(result[0]), Double.parseDouble(result[1]));
            }
            else {
                // reversed order
                origin = new Point2D.Double(Double.parseDouble(result[1]), Double.parseDouble(result[0]));
            }

```
It is less lines of code, and it highlights better what is different between the two cases. But I can accept the current code.

**Review by marpet — APPROVED** [review]

File: snap-jp2-reader/src/main/java/org/esa/snap/jp2/reader/metadata/Jp2XmlMetadata.java (hunk 3 above)
Comment (oanahogoiu) [reply snap-jp2-reader/src/main/java/org/esa/snap/jp2/reader/metadata/Jp2XmlMetadata.java]: Ok, refactoring done :)

**Pull request merged**